pkg/types/csv_rbac.go

```diff
@@ -0,0 +1,263 @@
+package types
+
+import (
+	"fmt"
+	"sort"
+)
+
+type CsvPermissions struct {
+	ClusterPermissions []Permissions `json:"clusterPermissions"`
+	Permissions        []Permissions `json:"permissions"`
+}
+
+type Permissions struct {
+	ServiceAccount string `json:"serviceAccount"`
+	Rules          []Rule `json:"rules"`
+}
+
+type Rule struct {
+	name            string   // used in tests
+	ApiGroups       []string `json:"apiGroups"`
+	Resources       []string `json:"resources"`
+	Verbs           []string `json:"verbs"`
+	ResourceNames   []string `json:"resourceNames"`
+	NonResourceURLs []string `json:"nonResourceURLs"`
+}
+
+type operator string
+
+type permissionType string
+
+type RuleFilter struct {
+	PermissionType           permissionType
+	ApiGroupFilterObj        *FilterObj
+	ResourcesFilterObj       *FilterObj
+	VerbsFilterObj           *FilterObj
+	ResourceNamesFilterObj   *FilterObj
+	NonResourceURLsFilterObj *FilterObj
+}
+
+type FilterObj struct {
+	Args         []string
+	OperatorName operator
+}
+
+type filterFunc func(*Rule, RuleFilter) *Rule
+
+const (
+	apiGroupFilterType        = "apiGroupFilter"
+	resourcesFilterType       = "resourcesFilter"
+	verbsFilterType           = "verbsFilter"
+	resourceNamesFilterType   = "resourceNamesFilter"
+	nonResourceURLsFilterType = "nonResourceURLsFilter"
+)
+
+var (
+	InOperator           operator = "IN"
+	NotInOperator        operator = "NOT_IN"
+	EqualsOperator       operator = "EQUAL"
+	NotEqualOperator     operator = "NOT_EQUAL"
+	ExistsOperator       operator = "EXISTS"
+	DoesNotExistOperator operator = "DOES_NOT_EXIST"
+	AnyOperator          operator = "ANY"
+)
+
+var (
+	AllPermissionType        permissionType = "all"
+	NameSpacedPermissionType permissionType = "namespaced"
+	ClusterPermissionType    permissionType = "clusterScoped"
+)
+
+func genFilter(attrFetcher func(*Rule) []string, filterType string) filterFunc {
```

Comment:
I have to admit it took me a while to understand what is happening here.

1. We define an attr fetcher to return a field of `Rule` type
2. We retrieve a FilterObj of the global `RuleFilter` struct using the custom switch statement
3. We return a new filterFunc that eventually evals the args against the filterObj

I feel the design could be simplify with clear interfaces, and contracts between objects.

For example, the `FilterObj` global struct really feels like different objects that should respect the `RuleFilter` interface. E.g:

```go
type RuleFilter interface {
    Filter(rule *Rule, args []string, op operator) *Rule
}

type ApiGroupFilter struct {}

func (f *ApiGroupFilter) Filter(rule *Rule, args []string, op operator) *Rule {
    return eval(rule.ApiGroups, args, op)
}
```

But IDK im just writing this quickly, maybe our interface expert @Ajpantuso would have some ideas.

Also I would dissociate the `PermissionType` from the `FilterObj`. From a filter's perspective, it should not care what rules it receives, as long as they are rules. Some external obj should determine if we should pass `clusterPermissions` or `permissions` or both.

Reply:
@sugarraysam Sure, let me try to refactor it, thanks for the feedback.
> Also I would dissociate the PermissionType from the FilterObj
 Yeah now that you've written it down, it makes total sense. :D